### PR TITLE
Hive with signature + use Bzz address accross Bee

### PR DESCRIPTION
**Pull request by pradovic**
- BzzAddress with signature in Hive as well
- Use whole multiaddr instead of just peerID  ni proto messages and signing
- Addressbook uses `BzzAddress` instead of multiaddress now
TODO:
- Unify `BzzAddress` proto filies in `bzz` module and use it from `handshake` and `hive` proto files

### CODE REVIEW

File: pkg/hive/hive_test.go [hunk 1]
```diff
@@ -189,37 +204,33 @@ func expectOverlaysEventually(t *testing.T, exporter ab.Interface, wantOverlays
...
 	}
 
-	m, err := exporter.Multiaddresses()
+	m, err := exporter.Addresses()
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	t.Errorf("Multiaddresses got %v, want %v", m, wantMultiaddresses)
+	t.Errorf("Multiaddresses got %v, want %v", m, wantBzzAddresses)
```
Comment (janos) [inline pkg/hive/hive_test.go:233]: Multiaddresses should be renamed to Bzz addresses.

File: pkg/p2p/libp2p/libp2p.go [hunk 2]
```diff
@@ -155,7 +155,12 @@ func New(ctx context.Context, signer beecrypto.Signer, networkID uint64, overlay
 		return nil, fmt.Errorf("autonat: %w", err)
 	}
 
-	handshakeService, err := handshake.New(overlay, h.ID(), signer, networkID, o.Logger)
+	underlays, err := buildFullMAs(h.Addrs(), h.ID())
+	if err != nil {
+		return nil, fmt.Errorf("handshake service: %w", err)
```
Comment (janos) [inline pkg/p2p/libp2p/libp2p.go:160]: This error annotation is not related to the action that buildFullMAs does.

File: pkg/bzz/address.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+		Signature: signature,
+	}, nil
+}
+
+func (a *Address) Equal(b *Address) bool {
+	return a.Overlay.Equal(b.Overlay) && a.Underlay.Equal(b.Underlay) && bytes.Equal(a.Signature, b.Signature)
+}
+
+func (p *Address) MarshalJSON() ([]byte, error) {
+	v := struct {
```
[the remark this answers is not in the document]
Comment (pradovic) [reply pkg/bzz/address.go:73]: I think I copy pasted the code, did not want to change much, but I agree, I'll add this as well

File: pkg/p2p/libp2p/libp2p.go (hunk 2 above)
Comment (pradovic) [reply pkg/p2p/libp2p/libp2p.go:160]: Ah, I thought it might be in the scope of creating a handshake service, but you are right, it's should have a separate error, good catch!

**Comment by pradovic** [comment]
@janos, sorry it ended up a bit longer. It was due to my lack of planning, I kind of saw things along the way and kept on going. Could have probably split it into multiple PRs. I'll try to pay more attention next time! Thanks for the review!

File: pkg/addressbook/addressbook.go [hunk 4]
```diff
@@ -90,15 +85,15 @@ func (s *store) Overlays() (overlays []swarm.Address, err error) {
 	return overlays, nil
 }
 
-func (s *store) Multiaddresses() (multis []ma.Multiaddr, err error) {
+func (s *store) Addresses() (multis []bzz.Address, err error) {
```
Comment (acud) [inline pkg/addressbook/addressbook.go:88]: multis should be renamed as it is no longer just multiaddresses

File: pkg/bzz/address.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+	"github.com/ethersphere/bee/pkg/crypto"
+	"github.com/ethersphere/bee/pkg/swarm"
+
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+var ErrInvalidAddress = errors.New("invalid address")
+
+type Address struct {
```
Comment (acud) [inline pkg/bzz/address.go:22]: it would be nice to have a godoc on this type, with explanation on how what the signature represents

File: pkg/bzz/address.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+	}
+
+	return &Address{
+		Underlay:  underlay,
+		Overlay:   overlay,
+		Signature: signature,
+	}, nil
+}
+
+func Parse(underlay, overlay, signature []byte, networkID uint64) (*Address, error) {
```
Comment (acud) [inline pkg/bzz/address.go:49]: if package would be called `bzzaddr` i think it would make the usage of such a function more clear in terms of outcome

File: pkg/bzz/address.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+	p.Overlay = a
+
+	m, err := ma.NewMultiaddr(v.Underlay)
+	if err != nil {
+		return err
+	}
+
+	p.Underlay = m
+	p.Signature, err = base64.StdEncoding.DecodeString(v.Signature)
```
Comment (acud) [inline pkg/bzz/address.go:106]: i would like to raise a flag here: you're not verifying that the signature is actually correct here. so theoretically i could somehow persist a wrong signature, deserialize it and it would turn out as a valid address, as opposed to `Parse` or `New` that actually calculate the signature. I would consider checking it here too

File: pkg/debugapi/peer.go [hunk 8]
```diff
@@ -35,21 +35,6 @@ func (s *server) peerConnectHandler(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	err = s.Addressbook.Put(address, addr)
```
Comment (acud) [inline pkg/debugapi/peer.go:38]: did we decide to remove this functionality? i'm not following why is this removed

File: pkg/hive/hive.go [hunk 9]
```diff
@@ -127,18 +130,20 @@ func (s *Service) peersHandler(_ context.Context, peer p2p.Peer, stream p2p.Stre
 		return fmt.Errorf("close stream: %w", err)
 	}
 	for _, newPeer := range peersReq.Peers {
-		addr, err := ma.NewMultiaddr(newPeer.Underlay)
+		bzzAddress, err := bzz.Parse(newPeer.Underlay, newPeer.Overlay, newPeer.Signature, s.networkID)
 		if err != nil {
-			s.logger.Infof("Skipping peer in response %s: %w", newPeer, err)
+			s.logger.Warningf("Skipping peer in response %s: %w", newPeer, err)
```
Comment (acud) [inline pkg/hive/hive.go:135]: no need to capitalize log messages, this is not done elsewhere

File: pkg/kademlia/kademlia.go [hunk 10]
```diff
@@ -170,6 +170,8 @@ func (k *Kad) manage() {
 					po := uint8(swarm.Proximity(k.base.Bytes(), peerToRemove.Bytes()))
 					k.knownPeers.Remove(peerToRemove, po)
 				} else {
+					k.logger.Errorf("this happened: %v", err)
```
Comment (acud) [inline pkg/kademlia/kademlia.go:173]: remove

File: pkg/bzz/address.go (hunk 6 above)
Comment (janos) [reply pkg/bzz/address.go:49]: Or to call the function `ParseAddress`?

File: pkg/bzz/address.go [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+	Underlay  ma.Multiaddr
+	Overlay   swarm.Address
+	Signature []byte
+}
+
+type addressJSON struct {
+	Overlay   string
+	Underlay  string
+	Signature string
+}
```
Comment (janos) [inline pkg/bzz/address.go:32]: JSON struct tags missing.

File: pkg/p2p/libp2p/libp2p.go [hunk 12]
```diff
@@ -155,7 +155,12 @@ func New(ctx context.Context, signer beecrypto.Signer, networkID uint64, overlay
 		return nil, fmt.Errorf("autonat: %w", err)
 	}
 
-	handshakeService, err := handshake.New(overlay, h.ID(), signer, networkID, o.Logger)
+	underlays, err := buildFullMAs(h.Addrs(), h.ID())
+	if err != nil {
+		return nil, fmt.Errorf("could not build host multiaddresses: %w", err)
```
Comment (janos) [inline pkg/p2p/libp2p/libp2p.go:160]: We do not need so long message, it could be just `"build host multiaddresses: %w"`. "Could not" is redundant given that it is an error.

File: pkg/bzz/address.go (hunk 7 above)
Comment (pradovic) [reply pkg/bzz/address.go:106]: The reasons why I did not add this check here is
1. It would need a networkID to be provided as well for the check
2. I somehow see it similarly to `&BzzAddress{ ... }` cuz it is just marshalling, unmarhsalling json. 

I can add the networkID to the function parameters as well, but I am not sure this is much useful, as these functions are made for  purpose of persisting  in the addressbook.

File: pkg/bzz/address.go (hunk 11 above)
Comment (pradovic) [reply pkg/bzz/address.go]: Ah, I added it in a later commit, sorry :)

**Comment by janos** [comment]
There are some problems that I encountered while running this branch on the cluster.

Debug API /topology endpoint is returning:

```
{
  "message": "topology json marshal interface error",
  "code": 500
}
```

All nodes are connected only to the bootnode as per debug API /peers endpoint. Probably due to error: `time="2020-05-29T10:10:34Z" level=error msg="error connecting to peer aaa5b52ff78aa240d8623bc3ef213b5a55f16cf3386556cc148fb25f8b9e79f0: failed to dial : all dials failed\n  * [/ip4/127.0.0.1/udp/7070/quic] CRYPTO_ERROR: peer IDs don't match"`.

This is now visible on the janos namespace which uses janos/bee:latest image built from this branch.

**Comment by pradovic** [comment]
@janos good catch, fixed! topology driver was not provided to the `Option` constructor

**Comment by janos** [comment]
Thanks for solving the /topology endpoint problem. I have created a new cluster in janos namespace from the latest commit in this branch. Some nodes have disconnections with error messages:

```
time="2020-05-29T11:09:43Z" level=error msg="error connecting to peer 8fdd31745d04b2af1a1827846cb1e19550254109253e11c0cb53f63e3d91f8a2: failed to dial 16Uiu2HAmUs8ySBXKZM6RqQeWmMXu6WnxwsnHjc9HS4HC5AdKLDba: no good addresses"
```

This can be seen with `kubectl logs pods/bee-3 -n janos`.

**Comment by pradovic** [comment]
@janos , @acud I have reverted `libp2p` connect related changes as it will take some more work to make it work nice with kademliia. Should brejust `BzzAddress` related stuff & `Hive` in this PR now.

File: pkg/addressbook/addressbook.go (hunk 4 above)
Comment (janos) [reply pkg/addressbook/addressbook.go:88]: This comment is left.

File: pkg/bzz/address.go (hunk 7 above)
Comment (janos) [reply pkg/bzz/address.go:109]: > My opinion is, if we end up using this on many places and need the check, then I would make these functions with some factory-style receivers or something similar.

I agree.

File: pkg/hive/hive.go [hunk 13]
```diff
@@ -11,34 +11,35 @@ import (
...
-	ma "github.com/multiformats/go-multiaddr"
 )
 
 const (
 	protocolName    = "hive"
 	protocolVersion = "1.0.0"
 	peersStreamName = "peers"
 	messageTimeout  = 1 * time.Minute // maximum allowed time for a message to be read or written.
-	maxBatchSize    = 50
+	maxBatchSize    = 30
```
[the remark this answers is not in the document]
Comment (pradovic) [reply pkg/hive/hive.go:28]: I tested, and reduced until it started working, which was around 40 I belive, so I reduced to 30. Let's test and see. We can reduce more, or increase limits on readers/writers.

File: pkg/hive/hive.go [hunk 14]
```diff
@@ -127,18 +130,20 @@ func (s *Service) peersHandler(_ context.Context, peer p2p.Peer, stream p2p.Stre
 		return fmt.Errorf("close stream: %w", err)
 	}
 	for _, newPeer := range peersReq.Peers {
-		addr, err := ma.NewMultiaddr(newPeer.Underlay)
+		bzzAddress, err := bzz.ParseAddress(newPeer.Underlay, newPeer.Overlay, newPeer.Signature, s.networkID)
 		if err != nil {
-			s.logger.Infof("Skipping peer in response %s: %w", newPeer, err)
+			s.logger.Warningf("skipping peer in response %s: %w", newPeer, err)
```
[the remark this answers is not in the document]
Comment (pradovic) [reply pkg/hive/hive.go:135]: I agree, but I am not sure we should disconnect from this node unless we document that we are punishing such a behavior, as it would impose other implementations of client to pay attention and check the addresses they are sending. What I am saying, it seems to be that this should be part of the protocol then., but it is a valid point! @janos, @zelig, what do you guys think as well?

File: pkg/p2p/libp2p/libp2p.go [hunk 15]
```diff
@@ -287,19 +285,25 @@ func (s *Service) AddProtocol(p p2p.ProtocolSpec) (err error) {
 	return nil
 }
 
-func (s *Service) Addresses() (addrs []ma.Multiaddr, err error) {
+func (s *Service) Addresses() ([]ma.Multiaddr, error) {
+	return buildFullMAs(s.host.Addrs(), s.host.ID())
+}
+
+func buildFullMAs(addrs []ma.Multiaddr, peerID libp2ppeer.ID) ([]ma.Multiaddr, error) {
```
[the remark this answers is not in the document]
Comment (pradovic) [reply pkg/p2p/libp2p/libp2p.go:292]: It is building a valid multi-address from PR and list of multi-addresses. It is the code that was already used in `libp2p`, implemented by @janos I believe. @acud what would you suggest as the name that would be easier to read?

File: pkg/p2p/libp2p/libp2p.go [hunk 16]
```diff
@@ -155,7 +155,12 @@ func New(ctx context.Context, signer beecrypto.Signer, networkID uint64, overlay
 		return nil, fmt.Errorf("autonat: %w", err)
 	}
 
-	handshakeService, err := handshake.New(overlay, h.ID(), signer, networkID, o.Logger)
+	underlays, err := buildFullMAs(h.Addrs(), h.ID())
+	if err != nil {
+		return nil, fmt.Errorf("build host multiaddresses: %w", err)
+	}
+
+	handshakeService, err := handshake.New(overlay, underlays[0], signer, networkID, o.Logger)
```
[the remark this answers is not in the document]
Comment (pradovic) [reply pkg/p2p/libp2p/libp2p.go:163]: I need any address that is a valid underlay address. It is also one of the topics I wanted to discuss after this PR, how do we tackle multiple valid underlays, and which one do we send in handshake and hive.

File: pkg/p2p/libp2p/libp2p.go [hunk 17]
```diff
@@ -155,7 +155,12 @@ func New(ctx context.Context, signer beecrypto.Signer, networkID uint64, overlay
 		return nil, fmt.Errorf("autonat: %w", err)
 	}
 
-	handshakeService, err := handshake.New(overlay, h.ID(), signer, networkID, o.Logger)
+	underlays, err := buildFullMAs(h.Addrs(), h.ID())
```
[the remark this answers is not in the document]
Comment (pradovic) [reply pkg/p2p/libp2p/libp2p.go:158]: This line here is just to check the error

File: pkg/node/node.go [hunk 18]
```diff
@@ -350,11 +353,19 @@ func NewBee(o Options) (*Bee, error) {
...
+					logger.Errorf("connect to bootnode %s", a)
+					return
+				}
+
+				err = addressbook.Put(overlay, *bzzAddr)
 				if err != nil {
 					_ = p2ps.Disconnect(overlay)
 					logger.Debugf("addressbook error persisting %s %s: %v", a, overlay, err)
-					logger.Errorf("persisting node %s", a)
+					logger.Errorf("connect to bootnode %s", a)
```
[the remark this answers is not in the document]
Comment (pradovic) [reply pkg/node/node.go:368]: I feel that to the user, it's in the scope of failed connect, as it will result in failed connect attempt. If the error message is `persisting node`, I believe it is not really user friendly, so I left it in Debug, for debugging. But I can also switch back, let me know what you think.

File: pkg/hive/hive.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (acud) [reply pkg/hive/hive.go:135]: SGTM :+1:

File: pkg/p2p/libp2p/libp2p.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (acud) [reply pkg/p2p/libp2p/libp2p.go:163]: exactly. how do you know that the first multiaddress is not the one which is bound to the loopback device for example?

[the remark this answers is not in the document]
Comment (pradovic) [reply pkg/p2p/libp2p/libp2p.go:163]: I agree, we need to tackle this, also if there is no multiaddresses, I don't see a check at the moment, etc.. So I think we should talk about this, basically the question is in what way we support multiple underlays. I am not sure should we solve it now or talk about it and solve it in separate PR?

File: pkg/p2p/libp2p/libp2p.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (acud) [reply pkg/p2p/libp2p/libp2p.go:158]: wouldn't be clearer to just call the multiaddress parser on one address?

[the remark this answers is not in the document]
Comment (pradovic) [reply pkg/p2p/libp2p/libp2p.go]: I refactored a bit, take a look at the last commit and let me know what you think

**Comment by pradovic** [comment]
@acud  no no, I fixed it after you made the comment, all good! :) Thanks a lot!
